pennylane/workflow/_capture_qnode.py

```diff
@@ -181,128 +177,117 @@ def _get_shapes_for(*measurements, shots=None, num_device_wires=0, batch_shape=(
+@qnode_prim.def_impl
+def _(*args, qnode, shots, device, qnode_kwargs, qfunc_jaxpr, n_consts, batch_dims=None):
+    if shots != device.shots:
+        raise NotImplementedError(
+            "Overriding shots is not yet supported with the program capture execution."
         )
+    if qnode_kwargs["diff_method"] not in {"backprop", "best"}:
+        raise NotImplementedError("Only backpropagation derivatives are supported at this time.")
 
-    # pylint: disable=unused-argument
-    @qnode_prim.def_abstract_eval
-    def _(*args, qnode, shots, device, qnode_kwargs, qfunc_jaxpr, n_consts, batch_dims=None):
+    consts = args[:n_consts]
+    non_const_args = args[n_consts:]
 
-        mps = qfunc_jaxpr.outvars
+    if batch_dims is None:
+        return device.eval_jaxpr(qfunc_jaxpr, consts, *non_const_args)
+    return jax.vmap(partial(device.eval_jaxpr, qfunc_jaxpr, consts), batch_dims[n_consts:])(
+        *non_const_args
+    )
 
-        batch_shape = (
-            _get_batch_shape(args[n_consts:], batch_dims[n_consts:])
-            if batch_dims is not None
-            else ()
-        )
 
-        return _get_shapes_for(
-            *mps, shots=shots, num_device_wires=len(device.wires), batch_shape=batch_shape
-        )
+# pylint: disable=unused-argument
+@qnode_prim.def_abstract_eval
+def _(*args, qnode, shots, device, qnode_kwargs, qfunc_jaxpr, n_consts, batch_dims=None):
+
+    mps = qfunc_jaxpr.outvars
+
+    batch_shape = (
+        _get_batch_shape(args[n_consts:], batch_dims[n_consts:]) if batch_dims is not None else ()
+    )
+
+    return _get_shapes_for(
+        *mps, shots=shots, num_device_wires=len(device.wires), batch_shape=batch_shape
+    )
+
+
+# pylint: disable=too-many-arguments
+def _qnode_batching_rule(
+    batched_args,
+    batch_dims,
+    *,
+    qnode,
+    shots,
+    device,
+    qnode_kwargs,
+    qfunc_jaxpr,
+    n_consts,
+):
+    """
+    Batching rule for the ``qnode`` primitive.
+
+    This rule exploits the parameter broadcasting feature of the QNode to vectorize the circuit execution.
+    """
+
+    for idx, (arg, batch_dim) in enumerate(zip(batched_args, batch_dims)):
+
+        if _is_scalar_tensor(arg):
+            continue
+
+        # Regardless of their shape, jax.vmap automatically inserts `None` as the batch dimension for constants.
+        # However, if the constant is not a standard JAX type, the batch dimension is not inserted at all.
+        # How to handle this case is still an open question. For now, we raise a warning and give the user full flexibility.
+        if idx < n_consts:
+            warn(
+                f"Constant argument at index {idx} is not scalar. "
+                "This may lead to unintended behavior or wrong results if the argument is provided "
+                "using parameter broadcasting to a quantum operation that supports batching.",
+                UserWarning,
+            )
+
+        elif arg.size > 1 and batch_dim is None:
```

Comment:
Comment on this branch no longer needed?

Reply:
Just a rewriting from:
```
if x:

else:
    if y:
```
to:
```
if x:

elif y:
```
My IDE was complaining, so I decided to follow it's suggestion.

Reply:
oh it did remove a comment when implementing the suggestion.